### PR TITLE
Refactor transform blocks to use BaseBlock and dedicated module

### PR DESCRIPTION
**Pull request by eshwarprasadS**
## Summary

- Migrates `FlattenColumnsBlock`, `SelectorBlock` (renamed to `IndexBasedMapperBlock`), and `SetToMajorityValue` from monolithic `utilblocks.py` to new directory structure: `src/sdg_hub/blocks/transform/`
- Updates all blocks to inherit from `BaseBlock` with standardized Pydantic validation and interfaces
- Adds explicit `input_cols` and `output_cols` parameters while maintaining full backward compatibility with existing flows
- Implements comprehensive column validation with improved error handling using `MissingColumnError`

### Architecture Updates
- **New location**: `src/sdg_hub/blocks/transform/` module with dedicated files for each block
- **BaseBlock integration**: Now uses Pydantic validation, rich logging panels, automatic validation, and standardized error handling
- **Standardized interfaces**: All blocks now use consistent `input_cols`/`output_cols` pattern
- **Enhanced constructors**: Support both legacy style and new BaseBlock pattern with explicit column specifications
- **Registry updates**: Uses new BlockRegistry pattern with category and description metadata

### Block-Specific Changes
- **SelectorBlock → IndexBasedMapperBlock**: Redesigned with multi-column choice mapping capability
- **Enhanced validation**: All blocks now use Pydantic field validators and model validators
- **Improved error handling**: Better error messages and validation coverage
- **Comprehensive tests**: Updated test expectations for improved error types and validation

## Blocks Migrated

1. **FlattenColumnsBlock** → `flatten_columns.py` - Wide-to-long format transformation with pandas melt
2. **IndexBasedMapperBlock** (formerly SelectorBlock) → `index_based_mapper.py` - Multi-column value selection and mapping based on choice columns  
3. **SetToMajorityValue** → `set_majority_value.py` - Sets all column values to the most frequent value

### Basic Single-Column Selection
```python
from sdg_hub.blocks.transform import IndexBasedMapperBlock

# Select responses based on a single judge verdict
block = IndexBasedMapperBlock(
    block_name="response_selector",
    input_cols=["response_1", "response_2", "response_3", "verdict"],
    output_cols=["selected_response"],
    choice_map={
        "Assistant A": "response_1",
        "Assistant B": "response_2", 
        "Assistant C": "response_3"
    },
    choice_cols=["verdict"]
)
```

### Multi-Column Selection (New Feature)
```python
# Select responses based on multiple judge verdicts simultaneously
block = IndexBasedMapperBlock(
    block_name="multi_judge_selector",
    input_cols=["model_a_response", "model_b_response", "human_response", 
                "judge_1_choice", "judge_2_choice"],
    output_cols=["judge_1_selection", "judge_2_selection"],
    choice_map={
        "model_a": "model_a_response",
        "model_b": "model_b_response",
        "human": "human_response"
    },
    choice_cols=["judge_1_choice", "judge_2_choice"]
)
```

### YAML Flow Configuration
```yaml
- block_type: IndexBasedMapperBlock
  block_config:
    block_name: response_selector
    input_cols: ["response_1", "response_2", "response_3", "verdict"]
    output_cols: ["selected_response"]
    choice_map:
      "Assistant A": "response_1"
      "Assistant B": "response_2"
      "Assistant C": "response_3"
    choice_cols: ["verdict"]
```

## Backward Compatibility

✅ All existing flows using these transform blocks continue to work unchanged  
✅ Import compatibility maintained through updated `__init__.py` files  
✅ Registry registration preserved with same block names  
✅ All constructor parameters work identically  
✅ Legacy constructor patterns fully supported

## Test Plan

- [x] All existing tests pass (22/22 for transform blocks, 400/400 overall)
- [x] Updated test error expectations for improved `MissingColumnError` vs `KeyError` handling
- [x] Import compatibility verified across all test files
- [x] BaseBlock validation and logging integration tested
- [x] Edge cases covered (missing columns, empty datasets, invalid choices)
- [x] Multi-column selection scenarios tested

Addresses #226 and follows the transform blocks reorganization outlined in #223.

## Summary by CodeRabbit

* **New Features**
  * Introduced data transformation tools for datasets, including:
    * Flattening columns from wide to long format with customizable input and output columns.
    * Mapping values between columns based on choice columns with detailed validation.
    * Standardizing a column to its most frequent (majority) value.

* **Bug Fixes**
  * Enhanced error handling with specific exceptions for missing or invalid columns in transformation blocks.

* **Tests**
  * Added comprehensive tests covering new transformation features, validation rules, and error conditions.

### CODE REVIEW

File: src/sdg_hub/blocks/transform/flatten_columns.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    """Block for flattening multiple columns into a long format.
+
+    This block transforms a wide dataset format into a long format by melting
+    specified columns into rows, creating new variable and value columns.
+
+    Parameters
+    ----------
+    block_name : str
+        Name of the block.
+    var_cols : List[str]
```
Comment (shivchander) [inline src/sdg_hub/blocks/transform/flatten_columns.py]: could we use input_cols for this?

File: src/sdg_hub/blocks/transform/flatten_columns.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        List of column names to be melted into rows.
+    value_name : str
```
Comment (shivchander) [inline src/sdg_hub/blocks/transform/flatten_columns.py]: and this would be output_cols

File: src/sdg_hub/blocks/transform/flatten_columns.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+        # Validate var_cols are subset of input_cols if both specified
+        if isinstance(self.input_cols, list) and self.var_cols:
+            missing_cols = set(self.var_cols) - set(self.input_cols)
+            if missing_cols:
+                logger.warning(
+                    f"Variable columns {missing_cols} not found in input_cols {self.input_cols}"
+                )
+
+    def _validate(self, samples: Dataset) -> Dataset:
```
Comment (shivchander) [inline src/sdg_hub/blocks/transform/flatten_columns.py]: we dont need a validate method right? Baseblock handles it, if we are having some custom validation, we should overload the _validate_custom method

File: src/sdg_hub/blocks/transform/selector.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    This block uses a mapping dictionary to select values from one column and
+    store them in a new output column based on a choice column's value.
+
+    Parameters
+    ----------
+    block_name : str
+        Name of the block.
+    choice_map : Dict[str, str]
+        Dictionary mapping choice values to column names.
+    choice_col : str
```
Comment (shivchander) [inline src/sdg_hub/blocks/transform/selector.py]: This would be input_cols

File: src/sdg_hub/blocks/transform/selector.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        Name of the column containing choice values.
+    output_col : str
```
Comment (shivchander) [inline src/sdg_hub/blocks/transform/selector.py]: This is output_cols

File: src/sdg_hub/blocks/transform/selector.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+        Name of the column to store selected values.
+    input_cols : Optional[Union[str, List[str]]], optional
+        Input column specification. Should include choice_col and mapped columns.
+    output_cols : Optional[Union[str, List[str]]], optional
+        Output column specification. Defaults to output_col.
+    num_procs : int, optional
```
Comment (shivchander) [inline src/sdg_hub/blocks/transform/selector.py]: Can we drop num_procs

File: src/sdg_hub/blocks/transform/selector.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+        Number of processes for parallel processing, by default 8.
+    **batch_kwargs : Dict[str, Any]
```
Comment (shivchander) [inline src/sdg_hub/blocks/transform/selector.py]: we can drop this as well

File: src/sdg_hub/blocks/transform/selector.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+                    f"Choice column '{self.choice_col}' not found in input_cols {self.input_cols}"
+                )
+
+            missing_mapped_cols = set(choice_map.values()) - set(self.input_cols)
+            if missing_mapped_cols:
+                logger.warning(
+                    f"Mapped columns {missing_mapped_cols} not found in input_cols {self.input_cols}"
+                )
+
+    def _validate(self, samples: Dataset) -> Dataset:
```
Comment (shivchander) [inline src/sdg_hub/blocks/transform/selector.py]: same as before - do we need this?

File: src/sdg_hub/blocks/transform/set_majority_value.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    """Block for setting all values in a column to the most frequent value.
+
+    This block finds the most common value (mode) in a specified column and
+    replaces all values in that column with this majority value.
+
+    Parameters
+    ----------
+    block_name : str
+        Name of the block.
+    col_name : str
```
Comment (shivchander) [inline src/sdg_hub/blocks/transform/set_majority_value.py]: replace this with input_cols

File: src/sdg_hub/blocks/transform/set_majority_value.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+
+        self.col_name = col_name
+
+        # Validate that col_name is in input_cols if specified
+        if isinstance(self.input_cols, list) and self.col_name not in self.input_cols:
+            logger.warning(
+                f"Column '{self.col_name}' not found in input_cols {self.input_cols}"
+            )
+
+    def _validate(self, samples: Dataset) -> Dataset:
```
Comment (shivchander) [inline src/sdg_hub/blocks/transform/set_majority_value.py]: do we need this?

File: src/sdg_hub/blocks/utilblocks.py [hunk 11]
```diff
@@ -95,74 +94,6 @@ def generate(self, samples: Dataset) -> Dataset:
         return samples
```
Comment (shivchander) [inline src/sdg_hub/blocks/utilblocks.py]: Do we need to remove it now, can't we handle the migration later?

File: src/sdg_hub/blocks/transform/set_majority_value.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+            logger.warning(
+                f"SetToMajorityValue modifies columns in-place. "
+                f"Specified output_cols {self.output_cols} will be ignored."
+            )
+        self.output_cols = []
+        
+        # Set derived attributes
+        self.col_name = self.input_cols[0]  # Use first (and only) input column
+
+    def _validate_custom(self, samples: Dataset) -> None:
```
Comment (shivchander) [inline src/sdg_hub/blocks/transform/set_majority_value.py]: Do we need this? I think this will be captured by the baseblock validation
